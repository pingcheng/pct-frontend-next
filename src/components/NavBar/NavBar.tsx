"use client";

import { profile } from "@/data/profile";
import Link from "next/link";
import { usePathname } from "next/navigation";
import styles from "./style.module.css";
import { useState } from "react";

type MenuItem = {
  label: string;
  path: string;
  activePattern?: RegExp;
};

const menuItems: MenuItem[] = [
  {
    label: "Home",
    path: "/",
  },
  {
    label: "Portfolio",
    path: "/portfolio",
    activePattern: /\/portfolio.*/,
  },
  {
    label: "About me",
    path: "/about",
  },
];

export function NavBar() {
  const pathName = usePathname();
  const [isMobileMenuVisible, setIsMobileMenuVisible] = useState(false);

  return (
    <>
      <nav>
        <div className="fixed-container relative flex items-center justify-between h-16">
          {/* Mobile nav bar*/}
          <div className="absolute inset-y-0 left-0 flex items-center sm:hidden">
            <button
              onClick={() => setIsMobileMenuVisible(!isMobileMenuVisible)}
              className="inline-flex items-center justify-center p-2 rounded-md text-gray-400 hover:text-white hover:bg-gray-700 focus:outline-hidden focus:ring-2 focus:ring-inset focus:ring-white"
              aria-expanded="false"
              role="mobile-toggle"
            >
              <span className="sr-only">Open main menu</span>

              <svg
                className={`block h-6 w-6 ${isMobileMenuVisible ? "hidden" : "block"}`}
                xmlns="http://www.w3.org/2000/svg"
                fill="none"
                viewBox="0 0 24 24"
                stroke="currentColor"
                aria-hidden="true"
              >
                <path
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="2"
                  d="M4 6h16M4 12h16M4 18h16"
                />
              </svg>

              <svg
                className={`block h-6 w-6 ${isMobileMenuVisible ? "block" : "hidden"}`}
                xmlns="http://www.w3.org/2000/svg"
                fill="none"
                viewBox="0 0 24 24"
                stroke="currentColor"
                aria-hidden="true"
              >
                <path
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="2"
                  d="M6 18L18 6M6 6l12 12"
                />
              </svg>
            </button>
          </div>

          {/* Desktop nav bar */}
          <div className="flex-1 flex items-center justify-center sm:items-stretch sm:justify-start">
            <div className="shrink-0 flex items-center">
              <div
                className="font-bold text-primary"
                style={
                  {
                    borderTop: "3px solid var(--logo-line-color)",
                    borderBottom: "3px solid var(--logo-line-color)",
                  } as React.CSSProperties
                }
              >
                {profile.fullName}
              </div>

              <div className="hidden sm:block sm:ml-6">
                <div className="flex space-x-4">
                  {menuItems.map((item) => {
                    return (
                      <Link
                        key={item.path}
                        href={item.path}
                        className={`${styles.navItem} ${
                          isActive(item, pathName) ? styles["active"] : ""
                        }`}
                      >
                        {item.label}
                      </Link>
                    );
                  })}
                </div>
              </div>
            </div>
          </div>
        </div>

        {/* Mobile nav bar */}
        <div
          className={`sm:hidden ${isMobileMenuVisible ? "block" : "hidden"}`}
          role="mobile-nav-menu"
        >
          <div className="px-2 pt-2 pb-3 space-y-1">
            {menuItems.map((item) => {
              return (
                <Link
                  key={item.path}
                  href={item.path}
                  className={`${styles.navItem} ${
                    isActive(item, pathName) ? styles["active"] : ""
                  }`}
                >
                  {item.label}
                </Link>
              );
            })}
          </div>
        </div>
      </nav>
    </>
  );
}

function isActive(item: MenuItem, pathName: string): boolean {
  if (item.activePattern) {
    return item.activePattern.test(pathName);
  }
  return pathName === item.path;
}
