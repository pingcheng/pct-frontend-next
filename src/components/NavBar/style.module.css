@reference "../../app/globals.css";

.navItem {
  @apply block text-gray-300 px-3 py-2 rounded-md text-sm font-medium transition duration-300;
  @apply bg-gray-700 text-white;
}

.navItem.active {
  @apply bg-gray-900 text-white;
}

@screen md {
  .nav-item {
    @apply inline-block;
  }
}
