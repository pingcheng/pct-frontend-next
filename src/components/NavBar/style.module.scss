.navItem {
    @apply block text-gray-300 px-3 py-2 rounded-md text-sm font-medium smooth;
    @apply bg-gray-700 text-white;

    &.active {
        @apply bg-gray-900 text-white;
    }
}