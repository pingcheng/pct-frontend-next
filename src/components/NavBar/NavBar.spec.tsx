import { render, screen, fireEvent } from "@testing-library/react";
import { NavBar } from "@/components/NavBar/NavBar";

jest.mock("next/navigation", () => ({
  usePathname: jest.fn(),
}));

import { usePathname } from "next/navigation";

describe("test <NavBar />", () => {
  it("should match the snapshot", () => {
    const { container } = render(<NavBar />);
    expect(container).toMatchSnapshot();
  });

  it("should toggle the mobile menu when the button is clicked", () => {
    render(<NavBar />);
    const toggleButton = screen.getByRole("mobile-toggle");
    const menu = screen.getByRole("mobile-nav-menu");
    // Initially, the mobile menu should be collapsed (not expanded)
    expect(menu.className).not.toMatch(/expanded/);
    // Click to show
    fireEvent.click(toggleButton);
    expect(menu.className).toMatch(/expanded/);
    // Click again to hide
    fireEvent.click(toggleButton);
    expect(menu.className).not.toMatch(/expanded/);
  });

  it("should display the correct SVG icon based on mobile menu visibility", () => {
    render(<NavBar />);
    const toggleButton = screen.getByRole("mobile-toggle");
    // Initially, hamburger icon should be visible, close icon hidden
    const hamburgerIcon = toggleButton.querySelector("svg:nth-of-type(1)");
    const closeIcon = toggleButton.querySelector("svg:nth-of-type(2)");
    expect(hamburgerIcon).not.toHaveClass("hidden");
    expect(closeIcon).toHaveClass("hidden");

    // Click to show mobile menu (should show close icon, hide hamburger)
    fireEvent.click(toggleButton);
    expect(hamburgerIcon).toHaveClass("hidden");
    expect(closeIcon).not.toHaveClass("hidden");

    // Click again to hide mobile menu (should show hamburger icon, hide close)
    fireEvent.click(toggleButton);
    expect(hamburgerIcon).not.toHaveClass("hidden");
    expect(closeIcon).toHaveClass("hidden");
  });
});

describe("isActive logic", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should apply active class for item with activePattern when path matches", () => {
    (usePathname as jest.Mock).mockReturnValue("/portfolio/123");
    render(<NavBar />);
    const portfolioLinks = screen.getAllByText("Portfolio");
    portfolioLinks.forEach((link) => {
      expect(link.className).toContain("active");
    });
  });

  it("should apply active class for item without activePattern when path matches exactly", () => {
    (usePathname as jest.Mock).mockReturnValue("/about");
    render(<NavBar />);
    const aboutLinks = screen.getAllByText("About me");
    aboutLinks.forEach((link) => {
      expect(link.className).toContain("active");
    });
  });

  it("should not apply active class for item without activePattern when path does not match", () => {
    (usePathname as jest.Mock).mockReturnValue("/not-matching");
    render(<NavBar />);
    const aboutLinks = screen.getAllByText("About me");
    aboutLinks.forEach((link) => {
      expect(link.className).not.toContain("active");
    });
  });
});

describe("Mobile menu interaction", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should close the mobile menu when a menu item is clicked", () => {
    (usePathname as jest.Mock).mockReturnValue("/");
    render(<NavBar />);
    const toggleButton = screen.getByRole("mobile-toggle");
    const menu = screen.getByRole("mobile-nav-menu");
    // Open the mobile menu
    fireEvent.click(toggleButton);
    expect(menu.className).toMatch(/expanded/);
    // Click a menu item
    const homeLink = screen.getAllByText("Home")[1]; // mobile menu is rendered after desktop
    fireEvent.click(homeLink);
    // Menu should be closed
    expect(menu.className).not.toMatch(/expanded/);
  });
});

describe("Mobile menu dynamic max-height", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should set max-height to scrollHeight when opening and 0px when closing", () => {
    (usePathname as jest.Mock).mockReturnValue("/");
    render(<NavBar />);
    const toggleButton = screen.getByRole("mobile-toggle");
    const menu = screen.getByRole("mobile-nav-menu");
    // Mock scrollHeight
    Object.defineProperty(menu, "scrollHeight", {
      value: 123,
      configurable: true,
    });
    // Initially closed
    expect(menu.style.maxHeight).toBe("0px");
    // Open the menu
    fireEvent.click(toggleButton);
    expect(menu.style.maxHeight).toBe("123px");
    // Close the menu
    fireEvent.click(toggleButton);
    expect(menu.style.maxHeight).toBe("0px");
  });
});
