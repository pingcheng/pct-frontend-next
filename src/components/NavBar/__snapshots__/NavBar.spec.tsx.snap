// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test <NavBar /> should match the snapshot 1`] = `
<div>
  <nav>
    <div
      class="fixed-container relative flex items-center justify-between h-16"
    >
      <div
        class="absolute inset-y-0 left-0 flex items-center sm:hidden"
      >
        <button
          aria-expanded="false"
          class="inline-flex items-center justify-center p-2 rounded-md text-gray-400 hover:text-white hover:bg-gray-700 focus:outline-hidden focus:ring-2 focus:ring-inset focus:ring-white"
          role="mobile-toggle"
        >
          <span
            class="sr-only"
          >
            Open main menu
          </span>
          <svg
            aria-hidden="true"
            class="block h-6 w-6"
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M4 6h16M4 12h16M4 18h16"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
          <svg
            aria-hidden="true"
            class="hidden h-6 w-6"
            fill="none"
            stroke="currentColor"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 18L18 6M6 6l12 12"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </button>
      </div>
      <div
        class="flex-1 flex items-center justify-center sm:items-stretch sm:justify-start"
      >
        <div
          class="shrink-0 flex items-center"
        >
          <div
            class="font-bold text-primary"
          >
            Ping Cheng
          </div>
          <div
            class="hidden sm:block sm:ml-6"
          >
            <div
              class="flex space-x-4"
            >
              <a
                class="navItem "
                href="/"
              >
                Home
              </a>
              <a
                class="navItem "
                href="/portfolio"
              >
                Portfolio
              </a>
              <a
                class="navItem "
                href="/about"
              >
                About me
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="sm:hidden hidden"
      role="mobile-nav-menu"
    >
      <div
        class="px-2 pt-2 pb-3 space-y-1"
      >
        <a
          class="navItem "
          href="/"
        >
          Home
        </a>
        <a
          class="navItem "
          href="/portfolio"
        >
          Portfolio
        </a>
        <a
          class="navItem "
          href="/about"
        >
          About me
        </a>
      </div>
    </div>
  </nav>
</div>
`;
